src/main/scala/GeomDataReader.scala

```diff
@@ -0,0 +1,164 @@
+package geotrellis.sdg
+
+import geotrellis.proj4._
+import geotrellis.proj4.util._
+import geotrellis.vector.{Extent, Geometry}
+import geotrellis.vector.reproject._
+import geotrellis.vectortile._
+import geotrellis.layer._
+
+import org.locationtech.geomesa.spark.jts._
+
+import org.locationtech.jts.geom.{Geometry => JTSGeometry}
+
+import org.apache.commons.io.IOUtils
+
+import org.apache.spark.sql.{SQLContext, DataFrame}
+import org.apache.spark.sql.functions.{udf, explode, lit}
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+import java.net.URI
+import java.util.zip.GZIPInputStream
+
+
+object GeomDataReader {
+  def readAndFormat(
+    sqlContext: SQLContext,
+    targetPath: String,
+    countryCode: String,
+    transform: MapKeyTransform
+  ): DataFrame = {
+
+    // This DataFrame contains the vector tile data
+    val tileDataFrame: DataFrame =
+      sqlContext
+        .read
+        .format("jdbc")
+        .options(
+          Map(
+            "url" -> s"jdbc:sqlite:$targetPath",
+            "driver" -> "org.sqlite.JDBC",
+            "dbtable" -> "tiles"
+          )
+        ).load()
+
+    // The data is sotred as pbf files, which are gziped vectortiles.
+    // So we need to decompress the bytes in order to read them in.
+    val decompressBytes: (Array[Byte]) => Array[Byte] =
+      (compressedBytes: Array[Byte]) => {
+        val stream = new GZIPInputStream(new ByteArrayInputStream(compressedBytes))
+        val output = new ByteArrayOutputStream()
+
+        val result = {
+          IOUtils.copy(stream, output)
+          output.toByteArray
+        }
+
+        stream.close
+        output.close
+
+        result
+      }
+
+    val decompressBytesUDF = udf(decompressBytes)
+
+    val uncompressedTileData: DataFrame =
+      tileDataFrame.withColumn("uncompressed_data", decompressBytesUDF(tileDataFrame.col("tile_data")))
+
+    val produceData: (Array[Byte], Int, Int) => Array[(JTSGeometry, String, String, String)] =
+      (bytes: Array[Byte], col: Int, row: Int) => {
+        val tileExtent: Extent = transform(col, row)
+
+        val vectorTile = VectorTile.fromBytes(bytes, tileExtent)
+
+        vectorTile.toIterable.toArray.map { feat =>
+          val featureType = feat.data.get("@type").get.asInstanceOf[VString].value
+
+          val roadType =
+            feat.data.get("highway") match {
+              case Some(value) => value.asInstanceOf[VString].value
+              case None => null
+            }
+
+          val surfaceType =
+            feat.data.get("surface") match {
+              case Some(value) => value.asInstanceOf[VString].value
+              case None => null
+            }
+
+          (feat.geom, featureType, roadType, surfaceType)
+        }
+      }
+
+    val produceDataUDF = udf(produceData)
```

Comment:
I think inlining the `udf` is a common style and saves having both `produceData` and `produceDataUDF`
example: https://github.com/locationtech/rasterframes/blob/2961077a5c6faf832120005fef37bf01dc928b94/core/src/test/scala/examples/NDVI.scala#L52-L56